### PR TITLE
TIMX-18-marc-transform-additional-fields

### PR DESCRIPTION
**Pull request by ehanson8**
#### How can a reviewer manually see the effects of these changes?
Run the transform locally using the `alma-2022-10-11-full-extracted-records-to-index.xml` file in the dev S3 bucket, e.g.:

```
pipenv run transform -i <S3 URL for the above file> -o alma-transformed-records.json -s alma
```

#### What are the relevant tickets?
* https://mitlibraries.atlassian.net/browse/TIMX-18

#### Developer
- [x] All new ENV is documented in README
- [x] Stakeholder approval has been confirmed (or is not needed)

#### Code Reviewer
- [ ] The commit message is clear and follows our guidelines
      (not just this pull request message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The documentation has been updated or is unnecessary
- [ ] The changes have been verified
- [ ] New dependencies are appropriate or there were no changes

#### Includes new or updated dependencies?
NO

### CODE REVIEW

File: tests/fixtures/marc/marc_record_all_fields.xml [hunk 1]
```diff
@@ -18,6 +18,22 @@
...
+            <subfield code="a">Tran, Phong,</subfield>
+            <subfield code="e">composer,</subfield>
+            <subfield code="e">performer.</subfield>
+        </datafield>
+        <datafield tag="110" ind1="2" ind2=" ">
+            <subfield code="a">Ceeys (Musical group),</subfield>
+            <subfield code="e">composer,</subfield>
+            <subfield code="e">performer.</subfield>
+        </datafield>
+        <datafield tag="111" ind1="2" ind2=" ">
```
Comment (ehanson8) [inline tests/fixtures/marc/marc_record_all_fields.xml:31]: I’ve included a full 111 field from our local data, these (and 711s) seem to be formatted this way in almost all the examples I found so I’m not sure if we want to process this differently than the Go version. The punctuation makes it quirky

File: transmogrifier/helpers.py [hunk 2]
```diff
@@ -12,11 +12,26 @@
 from lxml import etree  # nosec B410
 from smart_open import open
 
+from transmogrifier.config import load_external_config
 from transmogrifier.models import TimdexRecord
 
 logger = logging.getLogger(__name__)
 
 
+def crosswalk_value(crosswalk_path: str, type_value: Optional[str]) -> Optional[str]:
```
Comment (ehanson8) [inline transmogrifier/helpers.py:21]: I’m adding this parameter and more specific crosswalk files given the likelihood of overlapping letter codes for other fields but let me know if you see a cleaner approach

File: tests/test_marc.py [hunk 3]
```diff
@@ -53,11 +53,55 @@ def test_marc_record_all_fields_transform_correctly():
...
+            timdex.Contributor(
+                value="Tran, Phong, composer, performer",
+                kind="Main Entry - Personal Name",
+            ),
+            timdex.Contributor(
+                value="Ceeys (Musical group), composer, performer",
+                kind="Main Entry - Corporate Name",
+            ),
+            timdex.Contributor(
+                value="Theory of Cryptography Conference 2008 : New York, N.Y",
```
Comment (ehanson8) [inline tests/test_marc.py:77]: `rstrip` does have the side effect of removing abbreviation punctuation in some cases

**Review by hakbailey — COMMENTED** [review]
See inline comments for suggestions. I think the contributors field in particular would be a lot more similar to the records from our other sources with a few small changes.

File: transmogrifier/sources/marc.py [hunk 4]
```diff
@@ -89,13 +92,73 @@ def get_optional_fields(self, xml: Tag) -> dict:
...
+                datafield, "agrt", " "
+            ):
+                fields.setdefault("contents", []).append(contents_value)
 
         # contributors
+        contributor_marc_fields = [
+            {
+                "tag": "100",
+                "subfields": "abcdeq",
+                "kind": "Main Entry - Personal Name",
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply transmogrifier/sources/marc.py:117]: Agreed, I was having second thoughts about this as I was looking ahead to other fields. I'll pull in the relator and probably also remove the `Main Entry` and `Added Entry` language from alt title

File: transmogrifier/sources/marc.py [hunk 5]
```diff
@@ -89,13 +92,73 @@ def get_optional_fields(self, xml: Tag) -> dict:
...
+            fields["content_type"] = [
+                crosswalk_value(
+                    "config/marc_content_type_crosswalk.json", leader.string[6:7]
+                )
+            ]
 
         # contents
+        for datafield in xml.find_all("datafield", tag="505"):
+            if contents_value := self.create_subfield_value_string_from_datafield(
+                datafield, "agrt", " "
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply transmogrifier/sources/marc.py:108]: So I'm planning to continue using `create_subfield_value_string_from_datafield` but run a `split` after to turn it into a list, does that seem correct?

File: tests/test_marc.py [hunk 6]
```diff
@@ -53,11 +53,55 @@ def test_marc_record_all_fields_transform_correctly():
             "Dewey Call Number 3",
         ],
         citation=(
-            "Célébration : 10 siècles de musique de noël. "
+            "Célébration : 10 siècles de musique de noël. Language material. "
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply tests/test_marc.py:56]: I agree, I will make a ticket

File: transmogrifier/sources/marc.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (hakbailey) [reply transmogrifier/sources/marc.py:108]: What will you split on? How will you make sure separate subfield tags are split and metadata with separators in a single subfield are also split?

[the remark this answers is not in the document]
Comment (ehanson8) [reply transmogrifier/sources/marc.py:108]: I obviously misinterpreted your comment, let me dig into the data so I better understand

[the remark this answers is not in the document]
Comment (hakbailey) [reply transmogrifier/sources/marc.py:108]: Let me know if this seems wrong to you, but I think we want each subfield code entry to be its own item in the list unless the given subfield entry has the ` -- ` separator, in which case each item between the separator should be its own item in the list. We need to allow for the possibility of either case.

[the remark this answers is not in the document]
Comment (ehanson8) [reply transmogrifier/sources/marc.py:108]: This all makes sense, I'm tracking now. Mondays 🙃

File: tests/test_marc.py (hunk 3 above)
Comment (ehanson8) [reply tests/test_marc.py:77]: I redid the contributor `rstrip`, I don't think it's possible to clean up the `711`s without messing up the `700`s and `710`s so I'm prioritizing the formatting of those 2 given the infrequency of the `711`. I also added `rstrip` to `contents` which I think makes it more readable but let me know your thoughts

File: tests/test_marc.py [hunk 7]
```diff
@@ -53,11 +53,62 @@ def test_marc_record_all_fields_transform_correctly():
...
+            "Richard Wagner",
+            "(20:45)",
+            "Rückert-Lieder",
+            "Gustav Mahler",
+            "(19:33)",
+        ],
+        contributors=[
+            timdex.Contributor(
+                value="Tran, Phong",
+                kind="composer, performer",
```
[the remark this answers is not in the document]
Comment (hakbailey) [reply tests/test_marc.py:77]: We're also going to need to think about deduping contributors, right now there are a lot of records where the same contributor (with the same `value` and `kind`) is listed more than once.

**Review by hakbailey — CHANGES_REQUESTED** [review]
See a few suggestions/questions inline, once we get the contributors in line with our other sources this should be good to go.

File: transmogrifier/sources/marc.py [hunk 8]
```diff
@@ -22,12 +29,14 @@ def get_optional_fields(self, xml: Tag) -> dict:
...
 
+        leader = xml.find("leader", string=True)
+
         # alternate_titles
         alternate_title_marc_fields = [
             {
                 "tag": "130",
                 "subfields": "adfghklmnoprst",
-                "kind": "Main Entry - Uniform Title",
+                "kind": "Uniform Title",
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply transmogrifier/sources/marc.py:39]: I like it!

File: transmogrifier/sources/marc.py [hunk 9]
```diff
@@ -89,13 +98,81 @@ def get_optional_fields(self, xml: Tag) -> dict:
...
+        if leader:
+            fields["content_type"] = [
+                crosswalk_value(marc_content_type_crosswalk, leader.string[6:7])
+            ]
 
         # contents
+        for datafield in xml.find_all("datafield", tag="505"):
+            for contents_value in self.create_subfield_value_list_from_datafield(
+                datafield,
+                "agrt",
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply transmogrifier/sources/marc.py:113]: 👍

File: transmogrifier/helpers.py [hunk 10]
```diff
@@ -17,6 +17,21 @@
...
+    crosswalk_dict: Dict[Any, Any], value: Optional[str]
+) -> Optional[str]:
+    """
+    Crosswalk code to human-readable label based on the specified JSON crosswalk.
+
+    Args:
+        crosswalk_dict: A crosswalk dict of codes and replacement values.
+        value: A value to be crosswalked.
+    """
+    if value in crosswalk_dict:
```
[the remark this answers is not in the document]
Comment (ehanson8) [reply transmogrifier/helpers.py:30]: There is not!

File: tests/test_marc.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ehanson8) [reply tests/test_marc.py:77]: Updated the code to add separate contributors for each kind value and dedupe before setting the `fields` value. I tried to streamline a somewhat involved process as much as possible, see if I missed anything

**Review by hakbailey — APPROVED** [review]
Looks good!

**Pull request merged**